scitbx/lstbx/normal_equations.h

```diff
@@ -102,7 +103,7 @@ namespace scitbx { namespace lstbx { namespace normal_equations {
                     && b.size()   == w.size())(a.n_rows())(b.size())(w.size());
       SCITBX_ASSERT(a.n_cols() == n_parameters());
       sparse::matrix<scalar_t>
-      at_w_a = a.this_transpose_times_diagonal_times_this(w);
+      at_w_a = a.transpose().this_times_diagonal_times_this_transpose(w);
```

Comment:
After this pull request I've noticed that scaling in DIALS, when using Levenberg-Marquardt minimisation (which inherits from the normal_equations code), is around 15x slower. Reverting this line only seems to fix the issue. It looks like this line still does the same computation as before, just a low slower, and with reverting the lstbx tests and smtbx.refinement test still pass. Can this line be reverted or am I missing something regarding the behaviour here? Thanks, James

Reply:
Hi @luc-j-bourhis - the commit message says it should perform better :). Has anything else interfered with it?

Reply:
Thanks for pointing to the commit in question, hadn't checked this. If it's relevant, in dials we're typically only refining a few hundred parameters at most, maybe you are refining a lot more and the performance trade-offs change?

Reply:
I'm a little concerned the fact that commits from over three years ago has apparently been accidentally merged in means that we don't understand the changes that have been pulled in here?

Reply:
The change was definitively beneficial in my tests, which were all smtbx-centered. Could you point me toward the place where you use Levenberg-Marquardt in DIALS?

Reply:
@rjgildea came up with this to demonstrate:
```python
import time
from scitbx.array_family import flex
from scitbx.random import variate, uniform_distribution
from scitbx import sparse

m = 10000
n = 100
random_matrices = variate(
    sparse.matrix_distribution(
        m,
        n,
        density=0.6,
        elements=uniform_distribution(min=-3, max=10),
    )
)
w = flex.double_range(0, m)
ww = flex.double(m * m)
ww.resize(flex.grid(m, m))
ww.matrix_diagonal_set_in_place(diagonal=w)

a = random_matrices.next()
t0 = time.time()
for i in range(10):
    b = a.self_transpose_times_diagonal_times_self(w)
t1 = time.time()
print(t1 - t0)
for i in range(10):
    b = a.transpose().self_times_diagonal_times_self_transpose(
        w
    )
t2 = time.time()
print(t2 - t1)
```
though now I notice that the functions aren't necessarily expecting sparse matrices at all.

Reply:
The 'wider' the matrix becomes - the current implementation is faster. So for
`m = 100
n = 1000
`
the timing is approximately the same, for 
`m = 10
n = 1000
`
the timing figure get reversed to the original code snippet. I guess there should be a comparison of the matrix dimension to choose the implementation...

Reply:
Also, as @luc-j-bourhis mentioned - it does perform better for smtbx this way - the restraints matrix is much sparser that 0.6, try density=0.05 and a matrix with smtbx-suitable dimensions like 500x300 (500 parameters for 300 restraints) and also note that the current implementation scales much better then more restraints added. The easiest solution would be to add a bool option to control which sub-procedure is to use.

Reply:
Thanks for the insights, my cases are not usually very sparse and the matrix dimensions are quite different also. I agree we could add a bool option but this seems like a bit of a hack - I'd like to review my own code further and look further into what else is available in cctbx to see if I should be taking a different route. I'll make a separate issue to discuss this is necessary. Thanks, James

Reply:
If the sparsity of 0.6 is typical of your application in DIALS then I am pretty sure the overhead of my sparse matrix code means blunt dense code would do better. That member equation `add_equations` was designed for sparse problems only. Have you tried to use the `add_equation` member function? It lets you add one equation at a time, or said differently one row of the matrix you are currently passing to `add_equations` at a time.

Reply:
If that does not work for you, the clean way would be to add an overload to `add_equations` that takes a dense matrix. That would be really easy.

Reply:
Another way would be to add a function that takes a ready-to-use normal matrix of equations (e.g. computed with SciPy). This will then provide more flexibility regarding what functionality to use to get one...

Reply:
Thanks for the suggestions. To add further context to my recent pull request, I had since tried the converting to a dense matrix first as there is already an add_equations that takes a dense matrix. This was slightly quicker than using the new code, but still significantly slower than the old code as there is a loop over rows. I think it just happens that the way I have set up my problem is efficiently matched to the previous implementation, so having the option to switch to the other method of calculation is for now the easiest, although not the most elegant, solution.